Revive support for basic link alias

pderaaij opened the pull request:
## What are you trying to accomplish?
Implementing support of piped links to handle basic aliases of links. 

<img width="1145" alt="Screenshot 2021-05-25 at 20 51 07" src="https://user-images.githubusercontent.com/495374/119552560-f4867d80-bd9a-11eb-964e-e4d42f5381cd.png">


## What approach did you choose and why?
I've revived the implementation as in #223. It resembles the same functionality, but in the refactored code and with additional tests.

## Next steps

- There is an issues with piped links displayed as a table in Github Pages. This PR will not solve that problem. We should look into that specific issue.

pderaaij commented:
@riccardoferretti The change seems to work as it should. I am a bit sceptical about the tests. I needed to make sure the newly added test in `documents-link-provider.spec.ts` was the last one. Otherwise, I could run into flaky tests (see: https://github.com/foambubble/foam/runs/2668379484?check_suite_focus=true#step:7:174). Feels not good, but I can't imagine why the workspace is randomly disposed. Perhaps you have some ideas?

riccardoferretti reviewed (commented):
This looks really good, I would like to suggest another change, which is a big deeper but will make it much easier for us to (in a different PR) merge the model of `WikiLink` and `DirectLink`.

Basically in the future I would like to get to a point where I have this API:

```
interface Link {
  type: 'wikilink' | 'link'
  range: Range  // the range of the link (including brackets, to be checked whether now that's the behavior)
  target: string  // the target of the link (slug if wikilink, path if link)
  label: string // the label of the link (alias if wikilink with divider, slug if wikilink without divider, title if link)
  text: string // the original text from the document (what you call literalContent), including brackets (so it can support both links and wikilinks)
}
```

note that the fields will always exist (no `| null`), and have specific semantics.

if this feels a bit out of scope (which I can see why), then I'd say:
- rename `literalContent` to `text`, and have it include brackets
- rename `alias` to `label`, make it mandatory, and default to `slug` when no divider is present

I am also open to other names for `text`. Maybe `sourceText`, `rawText`, or let me know if you have other ideas now that I have shared with you some of the directional thinking for this model. (`literalContent` is not bad either, I wonder though if it can be "simpler")

pderaaij commented:
I took a first attempt at the refactor. If I understood it well, this should roughly resemble what you see for the future?

All functionalities still work, but an interesting thing became present. When I am not making use of the auto-generated markdown references the alias is properly shown in the markdown preview.

<img width="1300" alt="Screenshot 2021-05-27 at 20 40 04" src="https://user-images.githubusercontent.com/495374/119881660-d309cb00-bf2d-11eb-8d48-a7e364c8943d.png">

However, as soon as the references are generated the alias won't be shown in the preview any more. 
<img width="1221" alt="Screenshot 2021-05-27 at 20 40 31" src="https://user-images.githubusercontent.com/495374/119881726-ea48b880-bf2d-11eb-9a82-44afd3716240.png">

I need to look into the reason why this occurs. Perhaps good not to merge this PR before I've finished the investigation.

riccardoferretti commented on packages/foam-core/src/markdown-provider.ts:
```diff
@@ -190,10 +191,22 @@ const wikilinkPlugin: ParserPlugin = {
+        label: hasAlias
+          ? alias.trim()
+          : literalContent.substring(2, literalContent.length - 2),
+        target: (node.value as string).trim(),
```
should we use `data.permalink` here instead?

riccardoferretti commented on packages/foam-core/src/markdown-provider.ts:
```diff
@@ -422,7 +436,14 @@ export function createMarkdownReferences(
-      return { label: link.slug, url: pathToNote, title: target.title };
+      return {
+        label:
+          link.rawText.indexOf('[[') > -1
```
labels should only be created for wikilinks, this test should always be true, right?

riccardoferretti commented on packages/foam-core/src/markdown-provider.ts:
```diff
@@ -422,7 +436,14 @@ export function createMarkdownReferences(
+      return {
+        label:
+          link.rawText.indexOf('[[') > -1
+            ? link.rawText.substring(2, link.rawText.length - 2)
```
you raised a good point in your PR comment, I don't know what text goes here exactly, e.g. whether just the target or the raw text, we should look closely at the specs

pderaaij replied:
The format we use right now is the correct one if I follow the CommonMark specs, so suggest we stick to this.

riccardoferretti reviewed (commented):
the direction is definitely right.
I have left a few comments, let me know your thoughts.
thanks!

pderaaij commented:
I've been looking into the link aliasing with the link references. CommonMark does not allow for link aliases. The behaviour of CommonMark and GitHub flavour as such is to display the link with the pipe and the alias, e.g. `[link|alias]`. See this Gist as an example: https://gist.github.com/pderaaij/bd67bb6c0c0074389e70ef1249a8ba1e#file-test-alias-md.

As this is the behaviour of CommonMark the preview window of CommonMark displays the link fully as well. The target of the link is correct however, it's just not pretty.

The proper way to work with aliases in CommonMark is to use: `[integration][202105072028_Integration-in-organisation]`, see: https://gist.github.com/pderaaij/bd67bb6c0c0074389e70ef1249a8ba1e#file-test-only-alias-md

So, I see a couple of options moving forward:

1. implement link aliases for all situations, but knowing it would not work properly when using link references.
2. Only make link aliases work when the setting `foam.edit.linkReferenceDefinitions` is set to off. 
3. (in case of above) Create something for publishing to rewrite aliased links to CommonMark allowed format. But, don't know how to handle the preview screen in that case.

Or something else, but open for ideas here :)

riccardoferretti commented:
thanks for the investigation.
I would say we go for n. 3.
It will be possible to handle the preview by manually adjusting the links, see `preview-navigation.ts` - wanna take a stab at it?

pderaaij commented:
I'll have a look at it!

pderaaij commented on packages/foam-core/src/markdown-provider.ts:
```diff
@@ -190,10 +191,22 @@ const wikilinkPlugin: ParserPlugin = {
     if (node.type === 'wikiLink') {
+      const text = node.value as string;
+      const alias = node.data?.alias as string;
+      const literalContent = note.source.text.substring(
```
I am doing some tests with using `data.permalink` but that seems to break compatibility. The permalink fields will contained lowercased and hyphens (in place of spaces). If we moved forward to use this field I am assuming it breaks current setups that have spaces in their filenames. But, perhaps I am missing something here?

riccardoferretti replied:
oh good find, I wasn't aware of it. in this case I agree that we can't use the `permalink` property, but instead need to use the text from the wikilink (and strip the alias if present).

pderaaij commented:
I've been looking into it, but getting a bit stuck. So the code of `preview-navigation` is simple, but it seems not to be used. Even with all extensions disabled it doesn't use the code for the markdown preview. Probably I might be missing something here?

With regards to rewrite the alias I am bit searching what a right approach would be. As I see it right now foam has nothing present to manipulate files in publishing command or something like that? So, would this be something we'd like to add to the foam cli?

riccardoferretti commented:
> So the code of preview-navigation is simple, but it seems not to be used.

Not sure I understand, `preview-navigation.ts` is exposed in `features/index.ts` and then initialized with all the other features in `extension.ts` - is this what you are referring to?

> With regards to rewrite the alias I am bit searching what a right approach would be

if you look at `markdownItWithFoamLinks` in `preview-navigation.ts` you will see that we search for `[[wikilinks]]` and wrap them into a `<a href=...>wikilink</a>` for the preview html.
In this same context we could support the alias: `[[wikilink|alias]]` -> `<a href=...>alias</a>`

Is that helping?

pderaaij commented:
Yeah, I've been manipulating that piece of code. But even when I put in some dummy text it is not showing up in the preview. The debugger also doesn't reach that parrt of the code.. so, I am quite sure I am at the right part of the code. However, the code is not being used in my current setup and I don't know why.

riccardoferretti commented:
happy to look into this together, I think the work is really important and 98% there, we should push this through.

A couple of things that might help:
- I reviewed your code, you are based off an older version of master, that's why you don't see the decorating code being executed. Rebase from master to fix that
- I'd suggest you disable the wikilink definition generation while working on this, as it can add a level of complexity

Let me know if you want to pair up, or have questions, thanks again!

pderaaij commented:
The issue has to do with the link references. If they are present `markdown-it` will use these definitions to alter the state of the links. Rightfully so looking at the CommonMark spec. As such, I am manipulating the `markdown-it`plugin to disable the `reference` rule. That makes the link work, but leaves it up to me to remove the link definition block in the preview. Right now, I can only manipulate the AST tree, so need to find a better way for that. Hope to have some time to finish the last bit tomorrow.

Without the link definitions the preview is working properly.

pderaaij commented:
So, I woke up with a solution for this. It basically means we disable the default link rule of `markdown-it` and replace it with one that does not make use of references. Works for our use case, bust most likely interferes with other extensions. In any case, this does make the preview work accordingly. We could make it a configurable thing, so the choice is with the user?

pderaaij commented:
Okay, this is a bit weird. All the tests work on my [fork](https://github.com/pderaaij/foam/commit/e920999011f91c498184f65f58cf4ca663010c29), but not here as a PR. Any clue?

riccardoferretti commented:
it looks like that `link.target` in `markdown-provider.ts:132` is `undefined` or `null`, how could that happen?

pderaaij commented:
Okay, this is a bit weird. All the tests work on my [fork](https://github.com/pderaaij/foam/commit/e920999011f91c498184f65f58cf4ca663010c29), but not here as a PR. Any clue?

> it looks like that `link.target` in `markdown-provider.ts:132` is `undefined` or `null`, how could that happen?

Hmmz, weird that it only happens in a Windows environment in that case, and not in my personal fork? That feels off. I don't have the time to look into it tonight. Try to have a look tomorrow

riccardoferretti commented on packages/foam-core/src/plugins/index.ts:
```diff
@@ -0,0 +1,86 @@
+import * as fs from 'fs';
```
this file had been delete on master

riccardoferretti commented on packages/foam-vscode/src/features/preview-navigation.ts:
```diff
@@ -11,11 +12,14 @@ const feature: FoamFeature = {
-          md
-        ),
+      extendMarkdownIt: (md: markdownit) => {
+        // md.disable('reference');
```
remove

riccardoferretti commented on packages/foam-vscode/src/features/preview-navigation.ts:
```diff
@@ -29,13 +33,24 @@ export const markdownItWithFoamLinks = (
-        if (resource == null) {
-          return getPlaceholderLink(wikilink);
+        let linkHasAlias = wikilink.includes('|');
+        let resourceLink = wikilink;
+        if (linkHasAlias) {
+          resourceLink = wikilink.substring(0, wikilink.indexOf('|'));
+        }
```
minor: 

```suggestion
        const resourceLink = linkHasAlias  
          ? wikilink.substring(0, wikilink.indexOf('|'))
          : wikilink;
```

for two reasons:
1. let's use `const` whenever possible
2. this is easier to read

riccardoferretti commented on packages/foam-vscode/src/features/preview-navigation.ts:
```diff
@@ -29,13 +33,24 @@ export const markdownItWithFoamLinks = (
+          return getPlaceholderLink(resourceLink);
         }
+
+        let linkLabel = wikilink;
+        if (linkHasAlias) {
+          linkLabel = linkLabel.substr(wikilink.indexOf('|') + 1);
+        }
```
```suggestion
        const linkLabel = linkHasAlias
          ? linkLabel.substr(wikilink.indexOf('|') + 1)
          : wikilink;
```

riccardoferretti commented on packages/foam-vscode/src/features/preview-navigation.ts:
```diff
@@ -60,7 +75,7 @@ export const markdownItWithFoamTags = (
       try {
         const resource = workspace.find(tag);
-        if (resource == null) {
+        if (resource === null) {
```
the reason why we use `==` instead of `===` is because it also catches `undefined`, as opposed to just strictly `null`.

But the right approach here would be to use the utility function `isNone`

riccardoferretti commented on packages/foam-vscode/src/features/preview-navigation.ts:
```diff
@@ -77,4 +92,120 @@ export const markdownItWithFoamTags = (
+  return md.disable('link').use(linkRuleProcessingWithoutReferences());
+};
+
+function linkRuleProcessingWithoutReferences() {
```
I fear this approach is a bit heavy and disconnects us from future improvements in markdown-it.

Have you considered instead filtering the files to remove the section generated by Foam before it's sent to markdown-it? (not sure whether with the available API this is possible, have you already looked at that option?)

pderaaij replied:
I've tried to do that, but it seems manipulation can only be done on tokens. The link reference definition is not passed on as a token and as such manipulation is hard, The only thing I could find is to change the rule set of the inline link rules like I did above.

What still can be an option is that we don't display the alias when link references are present in the file. It then just shows `link|alias` in the preview. Which is actually the same as in GitHub without any manipulation of the Markdown files.

riccardoferretti replied:
mmm... I would need to look a bit more deeply into the API to properly help here.

I came across this: https://www.npmjs.com/package/markdown-it-replace-link - could it help?

pderaaij replied:
I would need to look at. The order in which all changes are executed seem to be of importance here. As soon as you save a document, the presentation of the link reference gets priority above the foam link regex. 

Fresh eyes probably help here ;-)

pderaaij replied:
For what I can see [vscode just reads the document ](https://github.com/microsoft/vscode/blob/77ff6eb03bcfb11d3bad26346f0b5ed593cb2a99/extensions/markdown-language-features/src/markdownEngine.ts#L132) and puts it in the parser of MarkdownIt. Then the tokenisation and rules start to apply. In both projects I don't see an possibility to intervene in this behaviour. For example, to filter out parts of the document before the tokenisation starts.

riccardoferretti reviewed (commented):
this is looking very good. I have left a couple of minor comments and one about the link definitions, let me know what you think

pderaaij commented:
Well, it has been quite a journey. I do believe an acceptable solution has been found. In this case, we add a custom rule to the link parser and iterate over the tokens to get the right behaviour. It was a bit of a journey, but it feels like the right approach. 

I fixed the latest remarks and squashed the commits. Please have a look at it @riccardoferretti and see what you think of it.

lautjy commented:
I understand that my comment is a bit late in the game, but do you have any thoughts on the order of the alias/wikilink?
Like in Dendron the order is [[alias | wikilink_338]], but here you are choosing reversed. Is there a reason.
Dendron docs on the matter: https://wiki.dendron.so/notes/3472226a-ff3c-432d-bf5d-10926f39f6c2.html

pderaaij commented:
@lautjy the WikiLink format is alias last. I actually came across an issue at Dendron to support the WikiLink format with alias last. So decided to take this order as it seemed the most common format.

riccardoferretti reviewed (approved):
LGTM!

StoltHD commented:
I have a problem with wikilinks with aliases in tables, those need to be escaped with a single backslash to work, but in Foam links with aliases that are escaped with backslash point to "notename\".
And if you add a link with alias without the "escape" the link do not work.
